api/types/volume/cluster_volume.go

```diff
@@ -0,0 +1,420 @@
+// alphanumeric character with '-', '_', '.', or alphanumerics in
+// between.
+type Topology struct {
+	Segments map[string]string `json:",omitempty"`
+}
+
+// CapacityRange describes the minimum and maximum capacity a volume should be
+// created with
+type CapacityRange struct {
+	// RequiredBytes specifies that a volume must be at least this big. The
+	// value of 0 indicates an unspecified minimum.
+	RequiredBytes uint64
+
+	// LimitBytes specifies that a volume must not be bigger than this. The
+	// value of 0 indicates an unspecified maximum
+	LimitBytes uint64
+}
+
+// Secret represents a Swarm Secret value that must be passed to the CSI
+// storage plugin when operating on this Volume. It represents one key-value
+// pair of possibly many.
+type Secret struct {
+	// Key is the name of the key of the key-value pair passed to the plugin.
+	Key string
+
+	// Secret is the swarm Secret object from which to read data. This can be a
+	// Secret name or ID. The Secret data is retrieved by Swarm and used as the
+	// value of the key-value pair passed to the plugin.
+	Secret string
+}
+
+// PublishState represents the state of a Volume as it pertains to its
+// use on a particular Node.
+type PublishState string
+
+const (
+	// StatePending indicates that the volume should be published on
+	// this node, but the call to ControllerPublishVolume has not been
+	// successfully completed yet and the result recorded by swarmkit.
+	StatePending PublishState = "pending-publish"
+
+	// StatePublished means the volume is published successfully to the node.
+	StatePublished PublishState = "published"
+
+	// StatePendingNodeUnpublish indicates that the Volume should be
+	// unpublished on the Node, and we're waiting for confirmation that it has
+	// done so.  After the Node has confirmed that the Volume has been
+	// unpublished, the state will move to StatePendingUnpublish.
+	StatePendingNodeUnpublish PublishState = "pending-node-unpublish"
+
+	// StatePendingUnpublish means the volume is still published to the node
+	// by the controller, awaiting the operation to unpublish it.
+	StatePendingUnpublish PublishState = "pending-controller-unpublish"
+)
+
+// PublishStatus represents the status of the volume as published to an
+// individual node
+type PublishStatus struct {
+	// NodeID is the ID of the swarm node this Volume is published to.
+	NodeID string `json:",omitempty"`
+
+	// State is the publish state of the volume.
+	State PublishState `json:",omitempty"`
+
+	// PublishContext is the PublishContext returned by the CSI plugin when
+	// a volume is published.
+	PublishContext map[string]string `json:",omitempty"`
+}
+
+// Info contains information about the Volume as a whole as provided by
+// the CSI storage plugin.
+type Info struct {
+	// CapacityBytes is the capacity of the volume in bytes. A value of 0
+	// indicates that the capacity is unknown.
+	CapacityBytes int `json:",omitempty"`
```

Comment:
I was looking at this, and initially wondering

- "should this be an `uint` / `uint64` if we don't use negative values
- would be consistent with `CapacityRange` (which uses `uint64`)
- ^^ could this be an issue? (are `CapacityRange` and `CapacityBytes` compared with each-other?)

it looks like the SwarmKit API type _does_ use an `int64`, so changing this type to the same would prevent having to do some type-juggling;

https://github.com/moby/moby/blob/1c129103b48f87b951bc2c4281feb7537c13ca41/vendor/github.com/moby/swarmkit/v2/api/types.pb.go#L4253-L4255


Looking slightly further, it looks like the SwarmKit API _does_ do some conversion to an `uint64`; https://github.com/moby/moby/blob/1c129103b48f87b951bc2c4281feb7537c13ca41/vendor/github.com/moby/swarmkit/v2/api/types.pb.go#L12860-L12868

And looking even further, it seems that for `CapacityRange` we do the exact oposite; the Swarm API uses `int64` but the moby type here uses `uint64`;
https://github.com/moby/moby/blob/1c129103b48f87b951bc2c4281feb7537c13ca41/vendor/github.com/moby/swarmkit/v2/api/types.pb.go#L4303-L4310

Perhaps we should align them, so that we don't have to do the type-juggling on conversion :thinking: